### PR TITLE
Dev/nem/mapd 359/nested markers

### PR DESCRIPTION
**Pull request by nmccready**
This is not to merge. A starting point for @zacronos or @tombatossals, to give me a hand.

A side note:
This library is well written but many functions are way too large and should be broken up into more readable functions. Also some helpers should be further broken down.

### CODE REVIEW

File: dist/angular-leaflet-directive.js [hunk 1]
```diff
@@ -1886,6 +1892,19 @@ angular.module("leaflet-directive").factory('leafletMarkersHelpers', ["$rootScop
         isObject = leafletHelpers.isObject,
         groups = {};
 
+   var _string = function(marker){
+       //this exists since JSON.stringify barfs on cyclic
```
Comment (nmccready) [inline dist/angular-leaflet-directive.js]: converting marker to a string easily for logging/debugging

File: dist/angular-leaflet-directive.js [hunk 2]
```diff
@@ -3736,14 +3771,17 @@ angular.module("leaflet-directive").directive('markers',
...
-                        {
-                            $it.each(newMarkers, function(markersToAdd){
-                                _addMarkers(markersToAdd, map, layers, leafletMarkers, leafletScope, shouldWatch);
+
+                        if(isNested) {
+                            $it.each(newMarkers, function(markersToAdd, layerName){
+                                _addMarkers(markersToAdd, map, layers, leafletMarkers, leafletScope, shouldWatch, layerName);
                             });
+                            console.debug('post');
+                            console.debug("layersMarkers equal: " + leafletMarkersHelpers.log(leafletMarkers.m1,true));
```
Comment (nmccready) [inline dist/angular-leaflet-directive.js]: this log statement shows layerMarkers. It also shows they were added to the map. Which is why I am confused why the test fails.

File: test/unit/layersDirective/layersDirective.overlays.markersSpec.coffee [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+                markers = leafletMarkers
+
+            scope.$digest()
+            leafletData.getLayers().then (layers) ->
+                leafletMarkersHelper.log markers.m1, true
+
+                expect(Object.keys(markers).length).toEqual 2
+                markerToCheck = markers.m1
+                expect(markerToCheck instanceof L.Marker).toBe true
+                expect(map.hasLayer(markerToCheck)).toBe true
```
Comment (nmccready) [inline test/unit/layersDirective/layersDirective.overlays.markersSpec.coffee]: test fails here

File: test/unit/layersDirective/layersDirective.overlays.markersSpec.coffee [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+            element = angular.element('<leaflet layers="layers" markers="markers" markers-nested="true"></leaflet>')
+            element = $compile(element)(scope)
+            map = undefined
+            leafletData.getMap().then (leafletMap) ->
+                map = leafletMap
+
+            markers = undefined
+            leafletData.getMarkers().then (leafletMarkers) ->
+                markers = leafletMarkers
```
Comment (nmccready) [inline test/unit/layersDirective/layersDirective.overlays.markersSpec.coffee]: usually missing _map which is odd.

File: dist/angular-leaflet-directive.js [hunk 5]
```diff
@@ -3736,14 +3771,17 @@ angular.module("leaflet-directive").directive('markers',
...
                             return;
                         }
                         _addMarkers(newMarkers, map, layers, leafletMarkers, leafletScope, shouldWatch);
```
Comment (zacronos) [inline dist/angular-leaflet-directive.js]: should this line be in an `else` branch?

Comment (nmccready) [inline dist/angular-leaflet-directive.js]: The return above in the if takes care of it.

Comment (zacronos) [inline dist/angular-leaflet-directive.js]: Ah, righto.

**Comment by nmccready** [comment]
@zacronos this is ready

**Comment by nmccready** [comment]
This does not include marker options handling yet. But this PR is big enough and we can do that next.

File: examples/overlays-markers-nested-example.html [hunk 6]
```diff
@@ -0,0 +1,140 @@
+<!DOCTYPE html>
```
Comment (nmccready) [inline examples/overlays-markers-nested-example.html]: nested example works now

File: grunt/concurrent.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
+'use strict';
+
```
Comment (nmccready) [inline grunt/concurrent.js]: allows tasks to run concurrently like watch and serve and specs and watch

File: test/unit/layersDirective/layersDirective.creationSpec.js [hunk 8]
```diff
@@ -0,0 +1,299 @@
+'use strict';
```
Comment (nmccready) [inline test/unit/layersDirective/layersDirective.creationSpec.js]: here down the layersDirective specs were broken up into separate files so I can read it. It was way too large.

**Comment by nmccready** [comment]
@tombatossals this is ready when it passes here I will merge to upstream.

**Comment by tombatossals** [comment]
LGTM :+1:

**Pull request merged**